src/zx/FunctionalityConstruction.cpp

```diff
@@ -0,0 +1,344 @@
+                    break;
+                }
+
+                case qc::OpType::Sdag: {
+                    addCphase(diag, PiRational(-1, 2), ctrl, target, qubit_vertices);
+                    break;
+                }
+
+                default: {
+                    throw ZXException("Unsupported Controlled Operation: " +
+                                      qc::toString(op->getType()));
+                }
+            }
+        } else if (op->getNcontrols() == 2) {
+            Qubit       ctrl_0 = 0;
+            Qubit       ctrl_1 = 0;
+            const Qubit target = op->getTargets().front();
+            int         i      = 0;
+            for (auto& ctrl: op->getControls()) {
+                if (i++ == 0)
+                    ctrl_0 = ctrl.qubit;
+                else
+                    ctrl_1 = ctrl.qubit;
+            }
+            switch (op->getType()) {
+                case qc::OpType::X: {
+                    addCcx(diag, ctrl_0, ctrl_1, target, qubit_vertices);
+                    break;
+                }
+
+                case qc::OpType::Z: {
+                    addZSpider(diag, target, qubit_vertices, Expression(),
+                               EdgeType::Hadamard);
+                    addCcx(diag, ctrl_0, ctrl_1, target, qubit_vertices);
+                    addZSpider(diag, target, qubit_vertices, Expression(),
+                               EdgeType::Hadamard);
+                    break;
+                }
+                default: {
+                    throw ZXException("Unsupported Multi-control operation: " +
+                                      qc::toString(op->getType()));
+                    break;
+                }
+            }
+        } else {
+            throw ZXException("Unsupported Multi-control operation (" + std::to_string(op->getNcontrols()) + " ctrls)" + qc::toString(op->getType()));
+        }
+        return it + 1;
+    }
+
+    ZXDiagram FunctionalityConstruction::buildFunctionality(const qc::QuantumComputation* qc) {
+        ZXDiagram           diag(qc->getNqubits());
+        std::vector<Vertex> qubit_vertices(qc->getNqubits());
+        for (size_t i = 0; i < qc->getNqubits(); ++i) {
+            diag.removeEdge(i, i + qc->getNqubits());
+            qubit_vertices[i] = i;
+        }
+
+        auto initial_layout     = qc->initialLayout;
+        auto output_permutation = qc->outputPermutation;
+
+        if (!initial_layout.empty()) {
+            std::vector<Vertex> new_qubit_vertices(qc->getNqubits());
+            for (auto i = 0; i < qc->getNqubits(); i++) {
+                new_qubit_vertices[i] = qubit_vertices[i];
+            }
+            for (auto& [tar, src]:
+                 qc->initialLayout) { // reverse initial permutation
+                if (tar == src)
+                    continue;
+
+                auto v_tar = diag.addVertex(tar, 1);
+                diag.addEdge(qubit_vertices[src], v_tar);
+                new_qubit_vertices[tar] = v_tar;
```

Comment:
Couldn't these lines also be covered pretty easily by employing a non-identity initial layout?